crates/shielder-relayer/src/price_feed.rs

```diff
@@ -0,0 +1,125 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use rust_decimal::Decimal;
+use serde::Deserialize;
+use strum::IntoEnumIterator as _;
+use strum_macros::EnumIter;
+use time::OffsetDateTime;
+use tokio::time::Duration;
+
+const BASE_PATH: &str = "https://api.diadata.org/v1/assetQuotation";
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, EnumIter)]
+pub enum Coin {
+    Eth,
+    Azero,
+    Btc,
+    Usdt,
+    Usdc,
+}
+
+/// A collection of prices for various coins.
+///
+/// The underlying structure is behind a mutex and a process to update it
+/// asynchronously can be strated with `start_price_feed`.
+#[derive(Clone)]
+pub struct Prices {
+    validity: time::Duration,
+    refresh_interval: Duration,
```

Comment:
why do we use two different `Duration`s? would it make sense to unify them?

Reply:
I imagine you'd want to set `refresh_interval` lower than `validity` (for example 1 min and 10 mins) so that:

1. There is no window where it's invalid but we have to wait to refresh
2. We get retries kind of "for free" - if an update fails, we just retry after the next `refresh_interval` and might still make it before `validity` elapses

Reply:
I meant - tokio's Duration and std's :)

Reply:
Ah, the API exposes the same `Duration`. One is more convenient internally, so I do the conversion in the constructor instead of on every usage. The conversion is just a long, painful line, so I prefer to do it this way.

Reply:
```
        let validity =
            time::Duration::new(validity.as_secs() as i64, validity.subsec_nanos() as i32);
```

yuck